### DISCUSSION TITLE
Beginner questions: How i use your lvgl_micropython on ESP32 S2 mini with ili9341

### DISCUSSION THREAD

**Discussion by fabse-hack** [lvgl-micropython/lvgl_micropython#75] [General]
Hello,

Yes, i read already your repo and your instructions how to build it.

i had some (beginner) questions:

here i manage the submodules?
https://github.com/kdschlosser/lvgl_micropython/blob/main/.gitmodules
for example: micropython, it automatic get the newest micropython rom from git? or i have to download my specific micropython version for the build?

How i config the hardware pins from the s2 mini to the ili9431 ? for example: pin8 go to CS ?

Can i use the square line studio (lvgl) for make an graphical design?

how is the process for using lvgl with your tool on my esp32 + display?
1. building the rom
2. flashing the rom
3. ????
4. ????

i thing lvgl is a very big tool for future projects, but right now, there are not many projects out there....

**Comment by kdschlosser** [comment]
You don't manage anything... It's all done internally in the build script for you...

It's as easy as cloning the repo and compiling. You need to have a C compiler installed.


```shell
git clone https://github.com/kdschlosser/lvgl_micropython
cd lvgl_micropython
python3 make.py esp32 submodules mpy_cross clean BOARD=ESP32_GENERIC_S3 BOARD_VARIANT=SPIRAM_OCT --flash-size=32 --oct-flash DISPLAY=ili9341 INDEV=gt911
```

As an example.

It will handle collecting MicroPython, LVGL and the ESP-IDF. It will set everything up and compile. You will need to have GCC, CMake, Ninja, Make, Git and Python 3.10+ installed prior to running the build script.

**Reply by fabse-hack** [reply]
ahhh thanks. I think, now i understand it better. For the s2 mini there are determined pins for the displays. I will try it 🤖

**Comment by fabse-hack** [comment]
Sorry, is me again 😅

I build now the firmware (is easy). And I flashed my esp32 s2 mini.

Can only connect to it, when I press 0 while plug it in....

And can you provide an example lvgl? How I make now a display design with buttons, widgets and so things?

Greetings

**Comment by kdschlosser** [comment]
Provide me a link to the actual board that you bought. I will look up the information about in on their website and get the information needed to get you up and running.

**Reply by fabse-hack** [reply]
https://amzn.eu/d/fFkQ2gf

s2 mini .... when i flash it with the normal micropthon, everything is normal, after flashing, the controller appear in the device manager. When i flash it with my lvgl builded version, then not .... :(

**Reply by fabse-hack** [reply]
...sorry for my late answer :(

**Reply by fabse-hack** [reply]
ah, for erase it ... i have to press 0 while plug-in the usb cable, then it appears in the device manager

**Comment by fabse-hack** [comment]
```
[1/1] cd /home/fabse/lvgl_micropython/lib/micropython/ports/esp32/build-LOLIN_S2_MINI-SPIRAM_OCT/bootloader/esp-idf/esptool_py && /home/fabse/.espressif/python_env/idf5.2_py3.12_env/bin/python /home/fabse/lvgl_micropython/lib/esp-idf/components/partition_table/check_sizes.py --offset 0x8000 bootloader 0x1000 /home/fabse/lvgl_micropython/lib/micropython/ports/esp32/build-LOLIN_S2_MINI-SPIRAM_OCT/bootloader/bootloader.bin
Bootloader binary size 0x48f0 bytes. 0x2710 bytes (35%) free.
[6/7] cd /home/fabse/lvgl_micropython/lib/micropython/ports/esp32/build-LOLIN_S2_MINI-SPIRAM_OCT/esp-idf/esptool_py && /home/fabse/.espressif/python_env/idf5.2_py3.12_env/bin/python /home/fabse/lvgl_micropython/lib/esp-idf/components/partition_table/check_sizes.py --offset 0x8000 partition --type app /home/fabse/lvgl_micropython/lib/micropython/ports/esp32/build-LOLIN_S2_MINI-SPIRAM_OCT/partition_table/partition-table.bin /home/fabse/lvgl_micropython/lib/micropython/ports/esp32/build-LOLIN_S2_MINI-SPIRAM_OCT/micropython.bin
micropython.bin binary size 0x224920 bytes. Smallest app partition is 0x225000 bytes. 0x6e0 bytes (0%) free.
Warning: The smallest app partition is nearly full (0% free space left)!
[7/7] Completed 'bootloader'
Project build complete. To flash, run:
idf.py flash
or
idf.py -p PORT flash
or
python -m esptool --chip esp32s2 -b 460800 --before default_reset --after no_reset write_flash --flash_mode dio --flash_size 4MB --flash_freq 80m 0x1000 build-LOLIN_S2_MINI-SPIRAM_OCT/bootloader/bootloader.bin 0x8000 build-LOLIN_S2_MINI-SPIRAM_OCT/partition_table/partition-table.bin 0x10000 build-LOLIN_S2_MINI-SPIRAM_OCT/micropython.bin
or from the "/home/fabse/lvgl_micropython/lib/micropython/ports/esp32/build-LOLIN_S2_MINI-SPIRAM_OCT" directory
python -m esptool --chip esp32s2 -b 460800 --before default_reset --after no_reset write_flash "@flash_args"
bootloader  @0x001000    18672  (   10000 remaining)
partitions  @0x008000     3072  (    1024 remaining)
application @0x010000  2246944  (    1760 remaining)
total                  2312480
make: Verzeichnis „/home/fabse/lvgl_micropython/lib/micropython/ports/esp32“ wird verlassen

Traceback (most recent call last):
  File "/home/fabse/lvgl_micropython/make.py", line 235, in <module>
    mod.compile(*extra_args)
  File "/home/fabse/lvgl_micropython/builder/esp32.py", line 1233, in compile
    raise RuntimeError('ERROR collecting esptool path')
RuntimeError: ERROR collecting esptool path
fabse@naotimi:~/lvgl_micropython$ python3 make.py esp32 submodules mpy_cross clean BOARD=LOLIN_S2_MINI BOARD_VARIANT=SPIRAM_OCT DISPLAY=ili9341 INDEV=gt911
```



now i can not build it anymore. Why the make.py want to flash it???? ahhhhh

**Reply by kdschlosser** [reply]
it isn't flashing it. The make.py script uses the flash tool to combine the 3 binaries the build makes into a single binary file.

**Reply by fabse-hack** [reply]
okay, thanks. I wonder already, why there is a bootloader.bin and micropython.bin. And now, my next wonder is: why the joining of the single files are failed .... ahhhh

hmmm, i will investigate it tomorrow and write down everything clear for me....

**Comment by kdschlosser** [comment]
if you run these commands it will flash. The build did succeed omnly the joining of the binary files into a single binary file failed...


```
cd /home/fabse/lvgl_micropython/lib/micropython/ports/esp32/build-LOLIN_S2_MINI-SPIRAM_OCT
/home/fabse/.espressif/python_env/idf5.2_py3.8_env/Scripts/python -m esptool --chip esp32s2 -b 460800 -p {PORT} --before default_reset --after no_reset write_flash "@flash_args"
```

replacing "{PORT}" with the port your ESP32 is attached to.

**Reply by fabse-hack** [reply]
yes, this is a "normal" flashing command. Okay, then your script gives me only the command. Thanks!

**Reply by kdschlosser** [reply]
I have to figure out why it is not completing. It should be able to locate the esptool path unless you have some kind of an odd setup going on.  I will take a look at it in a bit to figure out why it's having an issue.

**Reply by fabse-hack** [reply]
i found my mistake: I gave `BOARD_VARIANT=SPIRAM_OCT` to the ESP32 S2 .......... 

> if you run these commands it will flash. The build did succeed omnly the joining of the binary files into a single binary file failed...
> 
> ```
> cd /home/fabse/lvgl_micropython/lib/micropython/ports/esp32/build-LOLIN_S2_MINI-SPIRAM_OCT
> /home/fabse/.espressif/python_env/idf5.2_py3.8_env/Scripts/python -m esptool --chip esp32s2 -b 460800 -p {PORT} --before default_reset --after no_reset write_flash "@flash_args"
> ```
> 
> replacing "{PORT}" with the port your ESP32 is attached to.


`python.exe -m esptool --chip esp32s2 -b 460800 --port COM9 --before default_reset --after no_reset write_flash -z 0x1000 lvgl_micropy_LOLIN_S2_MINI-32.bin`

esptool want to have -z 0x1000 (i have it from the micropython flash instructions.......
but the esp32s2 mini not starting again ... something is wrong

```
Connected to MicroPython at com9
Use Ctrl-] or Ctrl-x to exit this shell


```

No REPL and I have to press "0" while plugin.... when i flash normale micropython on the same controller, everything is working.
I will test it now with another microcontroller...............

**Reply by fabse-hack** [reply]
so, i have an LILYGO S2_WOOR.
I make a firmware with the ESP32_GENERIC_S2. Flashing successfull. 
`python.exe -m esptool --chip esp32s2 -b 460800 --port COM9 --before default_reset --after no_reset write_flash -z 0x1000 lvgl_micropy_ESP32_GENERIC_S2-32.bin`

When i want to connect:
```
>python.exe -m mpremote connect com9
Connected to MicroPython at com9
Use Ctrl-] or Ctrl-x to exit this shell
r: 0xffffffff
invalid header: 0xffffffff
invalid header: 0xffffffff
invalid header: 0xffffffff
```

i think the write_flash -z 0x1000 is wrong ??????

**Reply by kdschlosser** [reply]
use thew write command I gave you. That is the one that the ESP-IDF spits out. my code doesn't change it.

**Reply by fabse-hack** [reply]
thanks. Yes, i think now, the builded firmware is working on my controller. But with mpremote i can not copy main.py to the controller.
`python.exe -m mpremote connect COM9 cp main.py :` is not working .... how i can copy some code to the micropython? or i have to build the firmware with code?

**Reply by kdschlosser** [reply]
personally I have never used `mpremote`. I use a highly modified version of `ampy` which is a python library. So there is not too much I am going to be able to help you with when using mpremote.

one day I will finish a program that I am working on that will create a fake removable drive on a PC allowing you to drag and drop files/folders directly to it. I had started working on it but ran into problems because of the crazy low speed of 115200 that MicroPython has set for the REPL. Now the ability to up the speed exists but in order to do it you have to upload a boot.py file that sets the speed so you still end up being forced to use some other script/program to transfer that file. I can write it into the program to handle that portion if I get to finish up writing the program. It mainly got stalled because of not being able to set the speed and I just haven't gotten the time to go back and finish it since the ability to adjust that was added.

I use the improved speeds with ampy and I am able to get almost 1mbit file transfers. It also transfers in parallel and my plan was to write into the firmware the code that specifically handles the transfer of files instead of having the additional overhead of passing python code along with the data that is being transferred.

**Reply by fabse-hack** [reply]
thanks. Is still not working. I try now 3 days to get some graphics on the ili9341

With my Lolin S2 mini, i can not connect to REPL --usb-otg is a unknown command for the LOLIN-S2-MINI.
The other controller that i have, lilygo esp32 s2 woor, i not found the right pin-outs for the ili9341....

i'm frustrated ............

//edit: for the Lolin S2 mini i use the command:

python3 make.py esp32 submodules mpy_cross clean BOARD=LOLIN_S2_MINI --flash-size=4 DISPLAY=ili9341 INDEV=gt911 --usb-otg

make: Unbekannte Option »--usb-otg«
unknow option...... without --usb-otg the REPL is not working....

**Reply by kdschlosser** [reply]
I have to check the MicroPython code for the LOLIN_S2_MINI to see how things are wired up. Give me a day to get that code correct.

It is only me that is working on this project and I do have a life outside of the computer. I am married, and that should about sum it up.. Gotta take time to spend it with the wife and I have things that I have to do around the house. Repairs and remodeling, things like that. The time that I spend writing code is usually at night from 7PM to 4AM or so and that's only when I am not able to sleep (which is 4 out of 5 days. Go to bed at 4:AM fall asleep around 5:00AM wake up at 8AM if i manage to fall asleep at all. Then it's work on the house and do other things like right now I am replacing the engine in a Subaru for a good friend of mine.

I understand your frustration and I am doing the best that I am able to with the time that I have.

**Comment by fabse-hack** [comment]
![image](https://github.com/user-attachments/assets/50751565-ee8e-4135-931c-7339b95ba3e4)

hahahahhaaaa 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀

**Comment by fabse-hack** [comment]
what lvgl version do you use in this project?

**Comment by fabse-hack** [comment]
![image](https://github.com/user-attachments/assets/d3858487-48fe-4449-89d3-b2f2898d3993)

@kdschlosser do you know SquareLine Studio?

![image](https://github.com/user-attachments/assets/b2cedef5-2c07-4cc9-afd5-22c6b09fce20)

**Comment by kdschlosser** [comment]
don't use squareline studio it doesn't write correct code for use with this binding.

**Reply by fabse-hack** [reply]
yes, i have to setup the display etc etc etc .... but for define the widget, is nice 2 have. I plan a project and take out the code snippets.

**Comment by fabse-hack** [comment]
https://github.com/fabse-hack/temp_humidity_micropython_lvgl/tree/main

is runable, but not yet finish :D

and i don't know if it's 100% correct. I call the lv.task_handler() a few times in some "while True" loops.

still have some questions, for exmaple: How i set the touch pins?

**Comment by kdschlosser** [comment]
what is the model of your touch IC?

**Reply by fabse-hack** [reply]
on the back is a small IC with the name XPT2046 2023

**Reply by fabse-hack** [reply]
Supported Touch IC's

CST816S
FT5x06
FT5x16
FT5x26
FT5x36
FT5x46
FT6x06
FT6x36
GT911
STMPE610
XPT2046


i'm make a binary file with -gt911 at the end. So this was wrong, i have to build it again with -xpt2046 at the end.

But how i define then the pins in the code?

**Reply by kdschlosser** [reply]
you have to add `INDEV=xpt2046` to the end of your build command..

I am guessing that is what you are using for a touch IC...

to add it to your program it depends on how your board is set up.. Does it share the same SPI bus as the display? or is it on it's own bus?

**Reply by kdschlosser** [reply]
what are the pins it is using?

**Reply by fabse-hack** [reply]
at the moment i use this pins:

```
spi_bus = SPI.Bus(
    host=1,
    mosi=11,
    miso=9,
    sck=7
)

display_bus = lcd_bus.SPIBus(
    spi_bus=spi_bus,
    dc=2,
    cs=15,
    freq=40000000,
)

display = ili9341.ILI9341(
    data_bus=display_bus,
    display_width=240,
    display_height=320,
    reset_pin=4,
    reset_state=ili9341.STATE_LOW,
    # backlight_pin=40,
    color_space=lv.COLOR_FORMAT.RGB565,
    # color_byte_order=ili9341.BYTE_ORDER_BGR,
    rgb565_byte_swap=True
)
```

so, is a good question, if this is on the same SPI.

![image](https://github.com/user-attachments/assets/f278e992-1c8e-4144-9652-e42b37836587)

so, on the right side they say TOUCH......

hmmmmmmmm, my plan is:

1. i will make a new binary file and flash it to the esp32 s2 mini
2. check the pins and the wiring of the display
3. testing with the same SPI (host=1)

**Reply by kdschlosser** [reply]
what display board are you using?

**Reply by fabse-hack** [reply]
ili9341 with xpt2046 IC

**Reply by kdschlosser** [reply]
who makes the display board?

**Reply by fabse-hack** [reply]
Hailege 2,8" ILI9341 240 * 320

is a china display......

**Reply by kdschlosser** [reply]
OK do you want it to use the same SPI pins as the display is using or do you want the touch to be on it's own set of pins.

I needed to see if the touch controller was broken out to it's own set of pins which is what I am seeing in the schematic for the display. 

The choice is yours as far as how you want to connect it. The SPI bus is able to be shared which saves on pins.

**Reply by fabse-hack** [reply]
hmmm okay, i will try to use this together? so SPI bus with sharing pins.

But i don't know, how the code have to look like. Can you provide me an "hello world" code with the touch pins and the useage with lvgl ?

that would be nice!
I will make a tutorial + youtube video, so maybe we can add my building examples in the readme.md on your repo ?
🚀 😄

**Comment by fabse-hack** [comment]
finish my project:

https://github.com/fabse-hack/temp_humidity_micropython_lvgl

... is a display for temperature & humidity, with time and MQTT.
@kdschlosser feel free to link it in your repo readme or not .... 🤓

**Reply by kdschlosser** [reply]
I will do just that...

One thing you might consider doing is packaging it all together... what I mean by that is you have the ability to add files to be frozen into the build and you can create a submodule in your repo for the binding repo. Write a simple makefile that will run the build and package everything into a single binary file. or optionally you can bake your source code into the firmware and release a binary for people to just download and flash .
